File: frontend/cypress/test-solutions/alien-dictionary.py
from collections import defaultdict, deque

class Solution:
    def alienOrder(self, words: list[str]) -> str:
        graph = defaultdict(set)
        in_degree = {char for word in words for char in word}
        
        for i in range(len(words) - 1):
            word1, word2 = words[i], words[i + 1]
            min_len = min(len(word1), len(word2))
            
            if len(word1) > len(word2) and word1[:min_len] == word2[:min_len]:
                return ""
            
            for j in range(min_len):
                if word1[j] != word2[j]:
                    if word2[j] not in graph[word1[j]]:
                        graph[word1[j]].add(word2[j])
                    break
        
        in_degree = {char: 0 for char in in_degree}
        for char in graph:
            for neighbor in graph[char]:
                in_degree[neighbor] += 1
        
        queue = deque([char for char in in_degree if in_degree[char] == 0])
        result = []
        
        while queue:
            char = queue.popleft()
            result.append(char)
            
            for neighbor in graph[char]:
                in_degree[neighbor] -= 1
                if in_degree[neighbor] == 0:
                    queue.append(neighbor)
        
        return "".join(result) if len(result) == len(in_degree) else ""

File: frontend/cypress/test-solutions/alien-dictionary.js
class Solution {
    alienOrder(words) {
        const graph = new Map();
        const inDegree = new Map();
        
        for (const word of words) {
            for (const char of word) {
                if (!graph.has(char)) {
                    graph.set(char, new Set());
                }
                if (!inDegree.has(char)) {
                    inDegree.set(char, 0);
                }
            }
        }
        
        for (let i = 0; i < words.length - 1; i++) {
            const word1 = words[i];
            const word2 = words[i + 1];
            const minLen = Math.min(word1.length, word2.length);
            
            if (word1.length > word2.length && word1.substring(0, minLen) === word2.substring(0, minLen)) {
                return "";
            }
            
            for (let j = 0; j < minLen; j++) {
                if (word1[j] !== word2[j]) {
                    if (!graph.get(word1[j]).has(word2[j])) {
                        graph.get(word1[j]).add(word2[j]);
                        inDegree.set(word2[j], inDegree.get(word2[j]) + 1);
                    }
                    break;
                }
            }
        }
        
        const queue = [];
        for (const [char, degree] of inDegree) {
            if (degree === 0) {
                queue.push(char);
            }
        }
        
        const result = [];
        while (queue.length > 0) {
            const char = queue.shift();
            result.push(char);
            
            for (const neighbor of graph.get(char)) {
                inDegree.set(neighbor, inDegree.get(neighbor) - 1);
                if (inDegree.get(neighbor) === 0) {
                    queue.push(neighbor);
                }
            }
        }
        
        return result.length === inDegree.size ? result.join('') : "";
    }
}